#42: update to be compatible with opentracing-api 0.30.0

jakerobb opened the pull request.

devinsba commented on src/main/java/brave/opentracing/BraveTracer.java:
```diff
@@ -52,7 +54,7 @@
  * @see Propagation
  */
-public final class BraveTracer implements Tracer {
+public final class BraveTracer extends ThreadLocalActiveSpanSource implements Tracer {
```
Brave already has its own active span source. There should be a wrapper around the brave abstraction so that spans created in OT are accessible in brave and the other way around

jakerobb replied:
Okay! I think I can make that happen.

jakerobb replied:
Followup question: is it reasonable/acceptable to expect that spans in a given application are all managed (activated and deactivated) either entirely using Brave or entirely using OT?

I've only got an hour into this so far, but it seems to me that OT's baked-in concept of a "stack" of spans, wherein the deactivation of one leads to the reactivation of whatever was active before it, will require that all activation/deactivation be done via the OT API.

jakerobb replied:
One idea to address that problem would be to add hooks in Brave itself to allow the ActiveSpanSource implementation to listen for changes to the CurrentTraceContext. Thoughts?

adriancole commented:
Maybe make a test for the problem? I think I get it, but that would help
steer things. This sort of thing would ideally have steered OT vs the other
way around, but we are where we are.

adriancole commented:
Feels like tracer impl should own the propagation impl as it knows how to
serialize a span (ex to get trace, span etc IDs for MDC etc). This should
be easiest to document and support even if later some sort of
synchronization efforts are made.

jakerobb commented:
What does "tracer impl should own the propagation impl" mean, exactly? Are you saying it should be Brave, and not the wrapping Tracer implementation, that should be responsible for the "stack" behavior? If that's the case, I don't think it's possible without changes to Brave.

Upon further investigation, it seems like Brave doesn't have a way to deactivate a span unless you have another span you want to make active. Am I missing something?

Also, the "stack" behavior I'm describing appears to be a bonus feature of ThreadLocalActiveSpanSource, rather than something required by the OT spec. Perhaps I should stop trying to retain the behavior.

jakerobb commented:
As for making a test for the problem -- I think what I'm asking here is whether there is, in fact, a problem.

adriancole commented:
>
> What does "tracer impl should own the propagation impl" mean, exactly? Are
> you saying it should be Brave, and not the wrapping Tracer implementation,
> that should be responsible for the "stack" behavior? If that's the case, I
> don't think it's possible without changes to Brave.
>
I need to look at the latest OT things to see what's been implied for
tracer maintainers wrt active span. Sounds like I don't understand it well
enough.

> Upon further investigation, it seems like Brave doesn't have a way to
> deactivate a span unless you have another span you want to make active. Am
> I missing something?
>
> In brave, you can clear the current span (temporarily or otherwise) vi
setting null in scope. But to clear a specific span from scope, you need
the reference to the handle returned when it was set in scope. there's no
management api or assumption that you can clear things you don't own.

> Also, the "stack" behavior I'm describing appears to be a bonus feature of
> ThreadLocalActiveSpanSource, rather than something required by the OT spec.
> Perhaps I should stop trying to retain the behavior.
>
+1

adriancole commented:
just looking at other impls, yeah I think best to not extend any types that aren't core. Ex lightstep's tracer adapts their span source into the active span api https://github.com/lightstep/lightstep-tracer-java/pull/94/files

bhs commented:
@jakerobb @adriancole if there are questions about intents / motivations from the OT / ActiveSpan standpoint, please `@bhs` me and I will do my best to help. I skimmed thread above and I think you've already realized that it's not *required* to do anything Stack-like.

jakerobb commented:
Thanks @bhs. The stack-like behavior seems extremely beneficial, though. Otherwise, everything that starts an in-process child span needs to take explicit steps to capture and reactivate the parent. It seems enormously beneficial to capture that behavior in the library, rather than have every instrumentation implementation do it (potentially incorrectly or inconsistently).

The fact that the built-in ThreadLocalActiveSpanSource has the stack behavior seems like a strong indicator that this is both correct and desirable, and that the thing to do is to update the documentation. JMHO. :)

jakerobb commented:
Peeling back another layer of the onion, I'm now learning that SpanInScope.close() provides stack-like behavior. This point might be moot. :)

jakerobb commented:
This should address Brian's requested changes. Committing before I leave for the day; I'll add some unit tests to cover the new classes tomorrow.

adriancole commented:
Thanks Jake. Will pull this down and test today with non-hypothetical
feedback ;)

adriancole commented on src/main/java/brave/opentracing/BraveActiveSpan.java:
```diff
@@ -0,0 +1,199 @@
+        return this;
+    }
+
+    public class Continuation implements ActiveSpan.Continuation {
```
this should be a static type, and likely doesn't need to be public since it doesn't expose anything not in the base type

jakerobb replied:
It's not static because it relies on several member variables of the parent class. The approach is copied from ThreadLocalActiveSpan in opentracing-util.

I can certainly add the requisite member variables and pass them in to Continuation's constructor if you feel it's important.

adriancole commented on src/main/java/brave/opentracing/BraveActiveSpan.java:
```diff
@@ -0,0 +1,199 @@
+import io.opentracing.Tracer;
+
+import java.util.Map;
+import java.util.Objects;
```
avoid adding java 8 library deps. if implementing eq/hashCode is needed, you can look at how some other brave types do that (usually it is via auto-value)

jakerobb replied:
I checked before adding this -- the POM specifies Java 8 for both source and target. Is that wrong?

adriancole commented on src/main/java/brave/opentracing/BraveActiveSpan.java:
```diff
@@ -0,0 +1,199 @@
+    }
+
+    @Override
+    public boolean equals(Object o) {
```
until there's a use case for eq/hashCode maybe best to leave it out. ex you'll notice this is sensitive to refCount and not sure this is a valuable

jakerobb replied:
Another remnant of a previous approach. Will remove.

adriancole commented on src/main/java/brave/opentracing/BraveActiveSpan.java:
```diff
@@ -0,0 +1,199 @@
+
+    @Override
+    public void deactivate() {
+        if (original) {
```
prefer early exit vs nesting on things like this.
ex `if (!original) return;`

adriancole commented on src/main/java/brave/opentracing/BraveActiveSpan.java:
```diff
@@ -0,0 +1,199 @@
+
+    @Override
+    public BraveActiveSpan setBaggageItem(String key, String value) {
+        wrapped.setBaggageItem(key, value);
```
is it possible to make a (package private) BraveBaseSpan (vs copy/pasting dispatch stuff which is easy to accidentally get wrong)

adriancole commented on src/main/java/brave/opentracing/BraveActiveSpan.java:
```diff
@@ -0,0 +1,199 @@
+ * @see BraveActiveSpanSource
+ * @see Tracer#activeSpan()
+ */
+public class BraveActiveSpan implements ActiveSpan {
```
final all the public types or hide their constructors as package-private

adriancole commented on src/main/java/brave/opentracing/BraveSpan.java:
```diff
@@ -25,7 +25,7 @@
  * <p>This type also includes hooks to integrate with the underlying {@linkplain brave.Tracer}.
  * Ex you can access the underlying span with {@link #unwrap}
  */
-public final class BraveSpan implements Span {
```
mistake? I don't see anything trying to extend BraveSpan..

ps I would make whatever subtype of basespan might eventually happen package-private and also prefer package private extension vs protected as the concrete ActiveSpan and Span would be in the same package as BaseSpan.

This sort of advice is mainly from effective java and permeates most of Brave code. Doing the same things here makes it easier to reason with, and prevents us from breaking users who use accidentally exposed things or accidentally not-final types.

jakerobb replied:
Tried to make BraveActiveSpan extend BraveSpan. Didn't end up liking the result and changed it back; forgot to clean this up. Will do!

adriancole commented on src/main/java/brave/opentracing/BraveSpanBuilder.java:
```diff
@@ -105,10 +111,32 @@
   /**
    * {@inheritDoc}
    */
-  @Override public BraveSpan start() {
```
the covariant override w/BraveSpan was there so that folks could unwrap into brave w/o casting.

adriancole commented on src/main/java/brave/opentracing/BraveTracer.java:
```diff
@@ -76,6 +94,11 @@ public static Builder newBuilder(Tracing brave4) {
       formatToPropagation.put(Format.Builtin.TEXT_MAP, brave4.propagation());
     }
 
+    public Builder setActiveSpanSource(ActiveSpanSource activeSpanSource) {
```
we don't use set prefix on things that are not javabeans (which means basically all code)

PS add javadoc on new public symbols at least linking to where to find more.

adriancole commented on src/main/java/brave/opentracing/BraveTracer.java:
```diff
@@ -76,6 +94,11 @@ public static Builder newBuilder(Tracing brave4) {
     }
 
+    public Builder setActiveSpanSource(ActiveSpanSource activeSpanSource) {
+      this.activeSpanSource = activeSpanSource;
```
null ok? is it ok to use any activeSpanSource?

jakerobb replied:
It should be okay. For example, ThreadLocalActiveSpanSource from opentracing-util works fine, as long as you don't care whether Brave's CurrentTraceContext knows about the current active span.

adriancole commented on src/main/java/brave/opentracing/BraveTracer.java:
```diff
@@ -105,23 +128,33 @@ public BraveTracer build() {
 
   BraveTracer(Builder b) {
     brave4 = b.brave4.tracer();
+    activeSpanSource = b.activeSpanSource;
```
if check null in builder, you can do this safely here as there's no other way to assign the value.

jakerobb commented:
Thanks for all the feedback. I'll address everything today.

adriancole commented:
>
> It should be okay. For example, ThreadLocalActiveSpanSource from
> opentracing-util works fine, as long as you don't care whether Brave's
> CurrentTraceContext knows about the current active span.
>
should add this to the javadoc, that using this option will prevent Brave's
underlying context integration from working. Personally, I'd rather leave
this out until requested, as it could create a support burden

adriancole commented:
>
> I checked before adding this -- the POM specifies Java 8 for both source
> and target. Is that wrong?
>
this is a trick. we use java 8 source and have retrolambda rewrite to java
7. However, we are java 7 compliant from a library POV (ideally would be
java 6, particularly as agents need that). So java 8 apis won't be
available on java 7 runtime. Surprised the build didn't fail on
animalsniffer.

adriancole commented:
>
> It's not static because it relies on several member variables of the parent class. The approach is copied from ThreadLocalActiveSpan in opentracing-util.
>
> I can certainly add the requisite member variables and pass them in to
> Continuation's constructor if you feel it's important.
>
no biggie

jakerobb commented:
Lots of things going on keeping me from making progress here, but I have addressed most (all?) of the code review feedback. Some unit tests around BraveActiveSpan and BraveActiveSpanSource are still needed.

I'm leaving on vacation until June 26th. I've added my coworker @dgrabows as a collaborator on my fork; further improvements may come from him while I'm gone.

We're currently struggling with an issue in which child spans appear as children of earlier siblings rather than their parent (A calls B, waits for B to return, and then calls C, but C gets recorded as a child of B). We were seeing it prior to upgrading opentracing-api to 0.30.0, so I don't think that's caused by anything in this library or my PR, but it's possible that additional changes and/or tests will arise as we get closer to a solution.

adriancole commented:
@jakerobb back from the holidays? 🌴 Seems 0.30 is still the latest opentracing version. Were there any follow-ups since your last note?

jakerobb commented:
Yep, thanks for the poke. 

I have a bit more work to do. At this point I'm reasonably confident that a change is needed in Brave itself in order to support the OpenTracing usage pattern. Specifically, Brave's two (default and Strict) implementations of CurrentTraceContext need to be modified to detect when the newScope method is called with a span that matches the span already held in their respective ThreadLocals, and when that happens, to return the existing scope rather than constructing a new one.

Without this change, usages of tracer.withSpanInScope can corrupt Brave's sense of the current span, such that deactivating the span leaves that same span still active despite it being finished, subsequently leading to traces that exhibit the behavior I described in my previous comment.

As I was coming up on an internal release deadline, we built a custom version of brave-opentracing that provides (and requires) its own CurrentTraceContext implementation. My intention is to migrate the change into Brave's native implementations, submit that as a separate PR, and then update this PR with corresponding changes to use those. I'll provide unit tests that demonstrate the problem as well.

Feedback appreciated on the plan.

adriancole commented:
>
> I have a bit more work to do. At this point I'm reasonably confident that
> a change is needed in Brave itself in order to support the OpenTracing
> usage pattern. Specifically, Brave's two (default and Strict)
> implementations of CurrentTraceContext need to be modified to detect when
> the newScope method is called with a span that matches the span already
> held in their respective ThreadLocals, and when that happens, to return the
> existing scope rather than constructing a new one.
>
the method newScope implies that a new scope object is being created.. I'm
not sure if we can guarantee all implementations (like grpc's Context or
Finagle's) would honor what you are trying to accomplish here.


> Without this change, usages of tracer.withSpanInScope can corrupt Brave's
> sense of the current span, such that deactivating the span leaves that same
> span still active despite it being finished, subsequently leading to traces
> that exhibit the behavior I described in my previous comment.
>
Brave's requires that if you call newScope, you call close on it. In
essence any impl should be provably correct even if the same span is
applied multiple times. If there's a path where this doesn't occur, then
this seems like a bug to me.

OpenTracing's conflation of scope and "finish" is a flaw pointed out at
least workshop. Hopefully, this can be remedied regardless of what we do
here (in python but same issue under ActiveSpan)
https://docs.google.com/document/d/1tSDlxOXnxWxl7v1FC-3r-CK15sknAWk42Ndkm5XkZ2U/edit#heading=h.zgj8rslf66dq

As I was coming up on an internal release deadline, we built a custom
> version of brave-opentracing that provides (and requires) its own
> CurrentTraceContext implementation. My intention is to migrate the change
> into Brave's native implementations, submit that as a separate PR, and then
> update this PR with corresponding changes to use those. I'll provide unit
> tests that demonstrate the problem as well.
>
yeah I'd like to get to the bottom of what is a flawed api design vs what
is a flaw in implementation. a test can help show that

adriancole commented:
@LotharSee I remember at the last workshop, the active span thing was discussed, and [many flaws were highlighted with its attempts to be automatic](https://docs.google.com/document/d/1tSDlxOXnxWxl7v1FC-3r-CK15sknAWk42Ndkm5XkZ2U/edit#). Is there any thing in progress to fix this before it ends up stuck in a lot of code?

Here's a refresher on the close scope ~ finish span thing:
* assumptions that of a particular scope would invoke close only once.
  * scope on its own can close idempotently
  * the penalty for calling close more than once is misaligns when you are counting to assume a  lifecycle
* there's only one thing that could close a scope
  * GRPC scope can close on a deadline, this could end early when you are counting to assume a lifecycle.

There are other glitches, like heavier responsibilities in the current design, but the most important one is that the scope+lifecycle counting thing is fragile. I don't think we should be investing too much in playing with this as it will lead folks to a cliff (and displace time along the way trying vainly to make it appear viable)

jakerobb commented:
All fair points; thanks for the context. I'll read through that thread and see what I can figure out. OT's use of reference counting is indeed fragile.

I moved the test down to Brave and ran it against latest code; it passes without any changes to the Tracer or CurrentTraceContext impls. Tried against 4.2.0 code (the version of the transitive dependency I'm getting from brave-opentracing), still passes. 🤦‍♂️  Problem must be in the bridge, but I sure as heck haven't been able to find it so far! Strong odds it's in my code. Will keep digging....

uditsaxena commented:
Hi, 

I'm currently using opentracing-api-0.30.0 with brave-opentracing. 

While 
`
braveTracer = BraveTracer.create(Tracing.newBuilder().sampler(CountingSampler.create(0.5f.toFloat)).reporter(new ZipkinR()).localServiceName("asdf").build())
`, 

I get the following error: 
```
Exception in thread "main" java.lang.NoSuchFieldError: SAMPLING_PRIORITY
	at brave.opentracing.BraveSpanBuilder.start(BraveSpanBuilder.java:119)
```

From the [link:](http://camel.465427.n5.nabble.com/CamelOpenTracing-broken-since-2-19-1-td5805709.html) I got to know that 
```
there was a breaking API change between OpenTracing 0.22 and 
0.30, and there is currently no released version of brave-opentracing that 
supports the OpenTracing 0.30.0 API. There is a PR in progress
```

which points to this PR. 

I wanted to confirm this is an issue and that it has been made visible. If so, I can move ahead and use the Brave implementation for now.
Are there previous versions which still work ? If so, which ones work ?

Thanks

bhs commented:
(see also https://github.com/opentracing/specification/issues/80 ... I will track this PR, too, but I wanted to put something under the OT github org to organize concerns about ActiveSpan since it's definitely quite important and consequential)

adriancole commented:
Hi, Jake. So do you have a failing test pushed at this point? If so I can
take a look tomorrow (also feel free to bump brave dep if you like)

Thanks for the progress

jakerobb commented:
@uditsaxena, that's not an "issue", per se, but a straightforward side effect of the fact that brave-opentracing-0.20.0 is not updated to use the new version of opentracing-api.

jakerobb commented:
@adriancole when I pushed the test down to Brave, it passed without changes. That led me to reanalyze my changes in brave-opentracing again, and for some inexplicable reason the thing that had eluded me for two weeks jumped right out and presented itself. I was effectively double-incrementing the refcount for a span any time you called makeActive. I've resolved that, and I just updated my branch/this PR with the fix.

jakerobb commented:
@adriancole I did the Brave upgrade to 4.5.0 as a separate PR: https://github.com/openzipkin/brave-opentracing/pull/45

adriancole commented on src/main/java/brave/opentracing/BraveActiveSpanSource.java:
```diff
@@ -0,0 +1,73 @@
+ * to do this, querying for the active span using the OpenTracing API will throw an exception.
+ */
+final class BraveActiveSpanSource implements ActiveSpanSource {
+  private final Map<Long, BraveActiveSpan> activeSpans = new ConcurrentHashMap<>();
```
It doesn't seems like the ActiveSpanSource contract requires returning exactly the same reference as prior calls. This cache should be documented I think..

jakerobb replied:
Will document. It's not in deference to the contract, but rather to the fact that Brave doesn't expose a way to get the current SpanInScope object. It's only available through calls to `withSpanInScope`, but you can't call that again because it will put another instance of the span on the stack, and we won't have an opportunity to close said span.

A change to Brave that exposed the current SpanInScope would greatly simplify this code. If you don't object, I can make that change as well and then revisit here.

adriancole replied:
The returned reference of the span in scope is usually a thread-scoped reference, but not always. Attempting to store a reference widens the implementation contract of brave (and census, and guice, and similar's) more than needed, because the safe usage pattern is always closing it on the same thread you opened it on.

For what you are trying to do here, it seems you are actually constrained to a thread-local implementation. For that, the ThreadLocal itself is a tool you could use. For example, you have a ThreadLocal here in this code, which catches the SpanInScope. you can check that thread local later.

jakerobb replied:
If I use my own ThreadLocal to track the active span, I wouldn't be able to stay in sync if an instrumenter used the OT API to make a span active, but then finished it with the Brave API. (That's actually how I tackled this in the first place, where my original impl used ot-util's ThreadLocalActiveSpanSource.)

The underlying thing here is that I need to depend on Brave to tell me what the active span is, but Brave doesn't offer the ability to access or close the scope associated with the current span unless you have access to the object that Brave returned when you made the span active in the first place.

adriancole commented on src/main/java/brave/opentracing/BraveActiveSpanSource.java:
```diff
@@ -0,0 +1,73 @@
+    BraveActiveSpan braveActiveSpan = activeSpans.get(spanId);
+    if (braveActiveSpan == null) {
+      braveActiveSpan = new BraveActiveSpan(this, tracer.withSpanInScope(span), BraveSpan.wrap(span));
+      activeSpans.put(spanId, braveActiveSpan);
```
there's a race condition here, which a concurrent test would point out. I'd push for the owner of the ActiveSpanSource interface to actually make a test suite so that all implementors don't need to copy tests here. This would also verify behaviors hidden but required cc @bhs @sjoerdtalsma

sjoerdtalsma replied:
I've been out of the tracing realm for a while unfortunately, and haven't been as involved with the OT API lately. I have no experience with the refcounting active span source myself.
@adriancole just to be clear; This is about OT 0.30 api, _not_ the java-spanmanager contrib library, right?

adriancole replied:
Yep testing impl against a contract. PS bug here is that loser should close its scope and return existing. However the impl still seems a bit off as scope is usually defined within a thread and the key here isnt thread scoped. If an api like this is to be retained, tests are really important and expensive to expect every impl to copy paste

jakerobb replied:
Continuations are an acknowledgement that a span might be started and finished on different threads. A thread-scoped key would make that impossible.

adriancole commented on src/main/java/brave/opentracing/BraveSpanBuilder.java:
```diff
@@ -105,10 +111,32 @@
+
+  @Override
+  public ActiveSpan startActive() {
+    if (!ignoreActiveSpan) {
```
this property itself hints maybe the ActiveSpan thing should be an extension, not a requirement

jakerobb replied:
That seems reasonable, but I think that having the library be responsible for keeping track of the active span greatly simplifies the application code for most use cases. I came into all of this as an application developer instrumenting services, and from that perspective I really want that stuff consolidated in one place.

adriancole commented on src/main/java/brave/opentracing/BraveTracer.java:
```diff
@@ -77,6 +95,19 @@ public static Builder newBuilder(Tracing brave4) {
+     * span to Brave and acts as a simple wrapper. You can override with any other implementation,
+     * but beware that some implementations, e.g. opentracing-util's ThreadLocalActiveSpanSource,
+     * may not tell Brave about the active span. In these scenarios, you would need to use the
+     * OpenTracing APIs exclusively, as the Brave APIs would not function correctly, if at all.
```
tests should be in place to show how this breaks.

adriancole commented on src/test/java/brave/opentracing/BraveTracerTest.java:
```diff
@@ -188,4 +196,80 @@ void checkSpanReportedToZipkin() {
+    assertEquals("SpanC's parent should be SpanA", idOfSpanA, parentIdOfSpanC);
+  }
+
+  @Test public void subsequentChildrenNestProperly_BraveStyle() {
```
a mixed test would be nice to show how they work together. Also unit tests especially for the racy code in package private classes so they don't bite later.

jakerobb commented on src/main/java/brave/opentracing/BraveActiveSpanSource.java:
```diff
@@ -0,0 +1,73 @@
+  }
+
+  private BraveActiveSpan getOrEstablishActiveSpan(brave.Span span) {
+    long spanId = span.context().spanId();
```
Will change. See previous comment about the requirement to use exact references.

adriancole commented:
>
> Continuations are an acknowledgement that a span might be started and
> finished on different threads. A thread-scoped key would make that
> impossible.
>
For behaviour like you are mentioning here and elsewhere, you need a lot
more tests. otherwise it will be easy to break

adriancole commented:
>
> The underlying thing here is that I need to depend on Brave to tell me
> what the active span is, but Brave doesn't offer the ability to access or
> close the scope associated with the current span unless you have access to
> the object that Brave returned when you made the span active in the first
> place.
>
The design of brave (also census) is that "you opened it, you close it". If
brave-native code opens a span scope, no one else should close that scope.
If you did, you would interfere with the calling code. In other words, you
should only close a scope you own, so if you originate a scope in
OpenTracing code, you should also close the scope there. The expectation
that you should be able to see a scope object created by upstream code is a
bit questionable in design, hence the other issue.

adriancole commented:
>
> If I use my own ThreadLocal to track the active span, I wouldn't be able
> to stay in sync if an instrumenter used the OT API to make a span active,
> but then finished it with the Brave API. (That's actually how I tackled
> this in the first place, where my original impl used ot-util's
> ThreadLocalActiveSpanSource.)
>
One thing about brave that might help is knowing that the scope is not
conflated with lifecycle at all. You are really placing a trace context in
scope (like finagle) not a span. Knowing this might help you think through
ideas. For example, you can't tell anything about lifecycle just by looking
at a span's scope.

What if you had a different tool. For example, what if brave had a finish
notification? Would this give you what you need here?

jakerobb commented:
Oh, that simplifies things greatly. I saw all the tests for mixed usage and assumed that it should be supported.

adriancole commented:
>
> Oh, that simplifies things greatly. I saw all the tests for mixed usage
> and assumed that it should be supported.
>
yeah I don't think I meant to show trading off on open/close, rather that
nesting itself works. I suspect I can clarify readme based on this chat.

jakerobb commented:
Okay, here's what I'm thinking:

* I'll make a class called UncloseableBraveActiveSpan (UBAS). It'll throw UnsupportedOperationException if you call close or deactivate on it, and continuations created from these will return UBAS's as well.
* ActiveSpanSource.activeSpan will return an instance of UBAS.
* Completely eliminate the cache of active spans.
* Document thoroughly.

I think this resolves all concerns, and eliminates the need for tests proving race conditions and other craziness.

adriancole commented:
>
> Okay, here's what I'm thinking:
>
>    - I'll make a class called UncloseableBraveActiveSpan (UBAS). It'll
>    throw UnsupportedOperationException if you call close or deactivate on it,
>    and continuations created from these will return UBAS's as well.
>    - ActiveSpanSource.activeSpan will return an instance of UBAS.
>    - Completely eliminate the cache of active spans.
>    - Document thoroughly.
>
> I think this resolves all concerns, and eliminates the need for tests
> proving race conditions and other craziness.
>
only note is maybe consider noop v throw (or some sort of counter or trace
log). Otherwise march on and keep us posted!

jakerobb commented:
Yeah, I had the same thought right after posting. ActiveSpans are AutoCloseables, so it's quite reasonable to assume people will call close() on them. I'm thinking throw on deactivate, noop on close.

jakerobb commented:
Okay, that came together easily, but it breaks a usage pattern I'm using, and one which I expect is common.

In my HTTP client library (it's a home grown wrapper for Ribbon), I have a `RequestInterceptor` interface with `beforeSubmit` and `onFinish` methods among others. I have an implementation of that interface, `OpenTracingRequestInterceptor`, which calls `startActive` in `beforeSubmit` and `deactivate` in `onFinish`. Since I'm not in the same method in which I called `startActive`, I'm using `tracer.activeSpan` to get the span to deactivate.

With the proposed changes, that call would throw UOE (or no-op, which isn't better). 

I could certainly modify my Ribbon library to have a "data bag" that it carries around with each request, and I could put the active span in there, but that only helps me. I think the ability to call `tracer.activeSpan().deactivate()` is required to fully support OT-api.

My next thought is to reintroduce the ThreadLocal, but that falls down for reasons discussed above -- it would only work if a service's instrumentation did 100% of its activating and deactivating with Brave or 100% with OT; no mixing allowed. Annotating/tagging/etc could still be mixed.

All of that leaves me with one final option: the one I was already pursuing. :)

adriancole commented:
>
> In my HTTP client library (it's a home grown wrapper for Ribbon), I have a
> RequestInterceptor interface with beforeSubmit and onFinish methods among
> others. I have an implementation of that interface,
> OpenTracingRequestInterceptor, which calls startActive in beforeSubmit
> and deactivate in onFinish. Since I'm not in the same method in which I
> called startActive, I'm using tracer.activeSpan to get the span to
> deactivate.
>
Stepping back here.. why do you need implicit "close counting" for
something like ribbon? RPC has defined start/stop/exception hooks. Why not
just stop the span when it is finished vs risking guessing wrong? There are
many instrumentation libraries that tackle this concern and it is pretty
library specific. At any rate, thanks for mentioning a specific use case,
as this will be relatively easy to compare magic vs explicit. I'll put a
gist up later

adriancole commented:
@jakerobb I suspect that your ribbon instrumentation isn't too much code to put as a throw-away commit. Mind doing that? One thing that would be nice, especially as this is use-case driver and open source, is to see what exactly you are doing (ex what version, which interface etc). Otherwise, I'd be guessing.

adriancole commented:
one guess at the description of `RequestInterceptor` I'm assuming you have one thread continuing between one callback and the next. eg it isn't actually switching threads between beforeSubmit and onFinish (as if it did, the context would have leaked). This is a fairly common scenario. For this scenario, you save a reference to the handle and close it on the other side.

https://github.com/openzipkin/brave/tree/master/brave#working-with-callbacks

If you have no means to pass the context handle between beforeSubmit and onFinish, but you know they are on the same thread, then the last resort is to define a thread local in `RequestInterceptor` to "hold it for you" between those calls.

It is very important to note that if there is a thread switch between `beforeSubmit` and `onFinish`, you should not set a span active, as it will leak the context onto the thread without a chance of cleanup. Doing that can create subtle bugs. If you are switching threads, then you should just pass a reference to the span itself

https://github.com/openzipkin/brave/tree/master/brave#working-with-callbacks-that-occur-on-different-threads

In general, my advice is that if you encounter a library that is difficult to instrument (eg pass scope on), open an issue on that library to try and make it easier. This saves you from having to instrument the thing with AOP or bytecode weaving.

jakerobb commented:
I think I can upload some sample code. It'll have to wait until I'm back in the office on Monday. 

In general, I think there's some good advice here. I'm also really encouraged by the direction of our other broader conversation. 🙂

jakerobb commented:
Took a PTO day yesterday to deal with a sick 3-year-old, but I'm back in action this morning.

I guess I'm not sure exactly what you're saying here:
> Stepping back here.. why do you need implicit "close counting" for
something like ribbon? RPC has defined start/stop/exception hooks. Why not
just stop the span when it is finished vs risking guessing wrong?

Ribbon handles some aspects of its lifecycle asynchronously, on different threads. Specifically, tracking of creation and reuse of HTTP connections from a pool is only detectable asynchronously. I'm annotating spans with those events. What will Brave do if I annotate a span after another thread has finished it?

> one guess at the description of RequestInterceptor I'm assuming you have one thread continuing between one callback and the next. eg it isn't actually switching threads between beforeSubmit and onFinish (as if it did, the context would have leaked). This is a fairly common scenario. For this scenario, you save a reference to the handle and close it on the other side.

I could indeed do that, but attempting to think as the library maintainer and not strictly about my own use case, it doesn't seem reasonable to force every user of the library to always track their own span references. It seems like tracer.activeSpan() is intended for exactly that purpose, and I'd be doing a disservice by crippling that. Do you disagree?

> In general, my advice is that if you encounter a library that is difficult to instrument (eg pass scope on), open an issue on that library to try and make it easier.

Agreed, but Netflix is no longer using or maintaining Ribbon. I don't think such a request would get traction. :(

jakerobb commented:
I put together a gist with a couple files I think will show how our Ribbon wrapper library is instrumented: https://gist.github.com/jakerobb/60151cf2482028f450fce5087829a1ae

adriancole commented:
Hi, Jake.

Thanks for writing this up, though I do think this is more something for
OpenTracing as I am certainly not the only person actively disinterested in
magic span closing being a part of implementing a tracer. That said I doubt
others will reply so I will. You did a good job covering points.

Ribbon handles some aspects of its lifecycle asynchronously, on different
threads. Specifically, tracking of creation and reuse of HTTP connections
from a pool is only detectable asynchronously. I'm annotating spans with
those events.

A request on a connection is a different scope than the connection itself.
As long as the eventa happen within the request response exception no prob.
You are hinting at tracking request independent activity which is
interesting but maybe not what others expect.

What will Brave do if I annotate a span after another thread has finished
it?

If you put late maintenance activity on the same span as a completed
request, it would be reported either by flush by you or a timeout. It isnt
normal to do this.

I could indeed do that, but attempting to think as the library maintainer
and not strictly about my own use case, it doesn't seem reasonable to force
every user of the library to always track their own span references. It
seems like tracer.activeSpan() is intended for exactly that purpose, and
I'd be doing a disservice by crippling that. Do you disagree?

I basically disagree that you have to conflate lifecycle to succeed in
tracing and also think it is a strange optimization for observability. We
have a difference of opinion here. It certainly isnt crippling, I dont
think others will call this crippling especially those who never had this
api

Agreed, but Netflix is no longer using or maintaining Ribbon. I don't think
such a request would get traction. :

Knowing is half the battle.. at the beginning of this thread there was no
context for the api, now ribbon, now getting more specific. Basically seems
worst case scenario.. we are reporting post request activity (not routinely
done and will surprise people and likely add confusion to what critical
path is) that occurs while using a dead library (a sad but true thing)

I've seen others poke at things, like modeling async maintenance work as
spans not annotation logs/spans. This would not break anything nor require
a special api. It might not be acceptable to you not sure. Point being
brainstorming on the problem of how to model ribbon could be something to
do before commiting to a tracer library requirement that then impacts
everyone in the ecosystem. Remember, this isnt just you and me, it is you
and me and instana and kamon and everybody else. We owe a lot of duty to
try to sort the problem itself imho, and I am grateful you gave details
about it as now we might be better able to.

jakerobb commented:
> A request on a connection is a different scope than the connection itself.
As long as the eventa happen within the request response exception no prob.
You are hinting at tracking request independent activity which is
interesting but maybe not what others expect.

To clarify:
I'm annotating spans with events that indicate which pooled connection was used for the request being traced. This lets me see how long Ribbon takes to acquire connections from its internal pool, which is interesting when comparing cold (empty/stale pool; connection must be established) and warm/hot (suitable connection already established) behavior.

![image](https://user-images.githubusercontent.com/8182938/28330790-d2b7f924-6bbc-11e7-8f41-3c06cde6e8f1.png)

jakerobb commented:
> we are reporting post request activity (not routinely
done and will surprise people and likely add confusion to what critical
path is)

It's parallel, not post. _Sometimes_ the last annotation comes after finish; sometimes it comes before (which is, in a way, worse than it being predictably post-request).

adriancole commented:
Thanks for clarifying. I am looking at another http model for tracing which
uses spans not logs/annotations for this sort of thing eliminating the
chance for this problem but also doesnt magically go in span detail pane at
the moment. It would be nice to eventually get some published models for
http lifecycle tracing as there are pros/cons and not much is discussed
generically so far.

On 19 Jul 2017 01:30, "Jake Robb" <notifications@github.com> wrote:

we are reporting post request activity (not routinely
done and will surprise people and likely add confusion to what critical
path is)

It's parallel, not post. *Sometimes* the last annotation comes after
finish; sometimes it comes before (which is, in a way, worse than it being
predictably post-request).

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/openzipkin/brave-opentracing/pull/43#issuecomment-316137469>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/AAD610_KrnWg0SHPkOz2zdAj9gJswUrRks5sPOu6gaJpZM4Nyv9V>
.

adriancole commented:
@devinsba if you have a minute, can you do a pass on this PR? Regardless of the future of ActiveSpan api ref counting etc, it is released in 0.30 so this PR implements that. If you are cool with this code, go ahead and merge.

@jakerobb on the backburner I will poke at ribbon, but currently eyes crossed in simplified json stuff. thx for the bread crumbs, and carry-on.

jakerobb commented:
@devinsba I have some small changes in flight to address Adrian's concerns about race conditions and unit test coverage.

richdougherty commented:
I built this branch locally because I needed an updated API version. I just wanted to let you know that this PR seemed to work fine for me in my simple tests. Thanks for implementing it!

I'm writing a coroutines implementation so in the end I used my own `ActiveSpanSource` so that I could suspend and resume stacks of `ActiveSpan`s during coroutine scheduling. I liked the fact that `BraveTracer`'s underlying `ActiveSpanSource` pluggable.

At first I was unsure about why there was a custom `BraveActiveSpanSource` instead of using OT's provided `ThreadLocalActiveSpanSource`, then I saw [this comment](https://github.com/openzipkin/brave-opentracing/pull/43/files/a7854242af256035800a31fc692571be7e5df9f2#r127232658):

> If I use my own ThreadLocal to track the active span, I wouldn't be able to stay in sync if an instrumenter used the OT API to make a span active, but then finished it with the Brave API. 

My question is, if I'm only using OT APIs, then is it fine to use something like the standard `ThreadLocalActiveSpanSource` instead of the custom `BraveActiveSpanSource` that you've written?

I see that the Brave APIs have their own parallel concept of a `SpanInScope`, but I believe I can just use my own `ThreadLocal` implementation to get an OT `ActiveSpan` and ignore Brave's `SpanInScope`. Is that right? I'm trying to implement suspendable/resumable `ActiveSpan` stacks and I'd prefer to build on the OpenTracing API if possible, just so it works across different frameworks.

Again, thanks for this PR and looking forward to seeing it merged in when it's finished. :)

adriancole commented:
> My question is, if I'm only using OT APIs, then is it fine to use something like the standard ThreadLocalActiveSpanSource instead of the custom BraveActiveSpanSource that you've written?

This is a good question. I don't think the benefits have been
thoroughly explained. Brave has a defined trace context data type,
which is not the case in OpenTracing at the moment. Due to this, there
are more things you can do. For example, brave's context can reliably
synchronize trace IDs into logs in the same way everywhere (SLF4J
etc), reuse other context storage like gRPC etc. When re-using
context-storage, it can leverage tooling built on that. For example,
in the case of grpc Context, this is used by Census. Census agent that
ensures grpc context is attached to uninstrumented thread pools would
magically work (without double-duty) with brave's CurrentTraceContext
impl. https://github.com/census-instrumentation/opencensus-java/pull/471

Choosing to use ThreadLocalActiveSpanSource eventhough you know you
are using brave undoes or waters down some of this value. For example,
there are duplicate projects in contrib for some of the things I've
mentioned above. However, they won't integrate as neatly as if you
literally know the contents of the trace context.

adriancole commented:
>
> I'm writing a coroutines implementation so in the end I used my own
> ActiveSpanSource so that I could suspend and resume stacks of ActiveSpans
> during coroutine scheduling. I liked the fact that BraveTracer's
> underlying ActiveSpanSource pluggable.
>
mind putting a gist up for some of your usage of this? There are precious
few practical applications and this is something we can change by
cataloging use in the real world

richdougherty commented:
OK I think I need to learn more about all these contexts. :) It sounds like Brave can be used with a `ThreadLocalActiveSpanSource`, but it won't have the advantages of using Brave's own context management.

I'm just getting started with integrating tracing into the coroutines scheduler, so it isn't quite working yet. I'll post up a Gist or link to the code when I get something working. Do you have a mailing list where I should post the link?

adriancole commented:
>
> I'm just getting started with integrating tracing into the coroutines
> scheduler, so it isn't quite working yet. I'll post up a Gist or link to
> the code when I get something working. Do you have a mailing list where I
> should post the link?
>
sure. you can post here https://groups.google.com/forum/#!forum/zipkin-user
or you can post into our gitter https://gitter.im/openzipkin/zipkin
or just reply :)

jakerobb commented:
> My question is, if I'm only using OT APIs, then is it fine to use something like the standard ThreadLocalActiveSpanSource instead of the custom BraveActiveSpanSource that you've written?

Yes, absolutely. In fact, in an early version of this PR, that's exactly what I was doing. It was decided that the benefits Adrian mentioned were significant enough that it made sense for this library to implement its own ActiveSpanSource.

adriancole commented:
Hey Jake. Can we please get some tests here that show things not working before opening issues on brave to change its api? There may simply be options to support OT 0.30 you havent thought of yet. Trying to guess is an ineffective and time consuming way for me to try to help. It would be easier and less distracting to help fix a failing test then guess.

Meanwhile please take watch of whatever is next in OT. A large amount of effort has been spent discussing how to deal with this api (and it continues still). If auto state management is removed or moved to a way that is easier to deal with, it can allow us to skip some of this work.

bhs commented:
@adriancole 

> If auto state management is removed or moved to a way that is easier to deal with, it can allow us to skip some of this work.

+1. @jakerobb https://github.com/opentracing/opentracing-java/pull/166 is not going to be merged as-is, but the discussion does seem to be trending pretty strongly towards an outcome where (at the very least) the refcounting will be pulled out of `Tracer` implementations. Would love to hear your thoughts over there.

jakerobb commented:
Sorry if I was too hasty in reporting an issue.

I'm not sure how to write a test to show that it's problematic that the bridge uses an internal cache of SpanInScopes keyed on SpanContext. Everything works as-is -- I just want it to be better. At a minimum, there's a potential memory leak, and a potential race condition, but if these are provable via unit test, I've been unable to do so.

The API feature I requested in Brave would have allowed me to eliminate the cache, and by extension any possibility of memory leak or race condition. If there are other ways to accomplish that, I'm open to ideas.

adriancole commented:
So brave has a type StrictCurrentTraceContext which can show leaks.
Something like this or home grown (unit test defined) on might help?

On 1 Aug 2017 4:53 am, "Jake Robb" <notifications@github.com> wrote:

> Sorry if I was too hasty in reporting an issue.
>
> I'm not sure how to write a test to show that it's problematic that the
> bridge uses an internal cache of SpanInScopes keyed on SpanContext.
> Everything works as-is -- I just want it to be better. At a minimum,
> there's a potential memory leak, and a potential race condition, but if
> these are provable via unit test, I've been unable to do so.
>
> The API feature I requested in Brave would have allowed me to eliminate
> the cache, and by extension any possibility of memory leak or race
> condition. If there are other ways to accomplish that, I'm open to ideas.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openzipkin/brave-opentracing/pull/43#issuecomment-319193121>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAD614ZeQ6BZcKR5GJH4v6ChFTZhRYI_ks5sTj68gaJpZM4Nyv9V>
> .
>

mpetazzoni commented:
FWIW I've been using this branch/fork with OT 0.30.0 and I'm not experiencing any major issues. I think this should be merged upstream to offer proper compatibility with OT 0.30.0. If there are any nits remaining, we can keep working on them afterwards.

adriancole commented:
I think the main problem is that there are no tests (besides manual testing
like you have done)

Probably makes sense to merge and have Jake raise his additional change
after the fact as this PR sitting has been revlocking folks regardless of
what happens in 0.31

adriancole commented:
I dont know how to summarize the problems with the active span source impl,
but these would effectively be the "release notes" until a patch fixes or
post factum test elaborates (or both)

mpetazzoni commented:
I must admit I got a little lost reading through the comments on this PR so I'm not exactly sure what the issues are. But as you said, this has been sitting for a while and is revlocking people. I think it could make sense to merge this and cut a release of `brave-opentracing`. Other issues discovered by more testing (manual, or automated, if more tests get written) can be addressed by further point-releases.

adriancole commented:
I'll choose the blue pill

adriancole commented:
pushing  release 0.21.0

@jakerobb if OT 0.31 doesn't obviate the active span problems, please do follow-up with a separate PR later, before we forget.

mpetazzoni commented:
Hmm, you cut the `release-0.21.0` tag but didn't update the POM version. Did you forget to `mvn release:prepare`?

adriancole commented:
the release-0.21.0 tag sets off automation to do the other things..

mpetazzoni commented:
Which maybe has failed? https://travis-ci.org/openzipkin/brave-opentracing/builds/260460516

adriancole commented:
yep. redirect to here
https://github.com/openzipkin/brave-opentracing/pull/47
